modules/storages/app/services/storages/one_drive_managed_folder_sync_service.rb

```diff
@@ -0,0 +1,193 @@
+    end
+
+    def hide_inactive_folders(folder_map)
+      project_folder_ids = active_project_storages_scope.pluck(:project_folder_id).compact
+      (folder_map.keys - project_folder_ids).each do |item_id|
+        Peripherals::Registry.resolve("commands.one_drive.set_permissions")
+                             .call(storage: @storage, path: item_id, permissions: { write: [], read: [] })
+                             .on_failure do |service_result|
+          format_and_log_error(service_result.errors, folder: path, context: 'hide_folder')
+        end
+      end
+    end
+
+    def add_user_to_permission_list(permissions, token, project)
+      op_user_permissions = token.user.all_permissions_for(project)
+
+      if op_user_permissions.member?(:write_files)
+        permissions[:write] << token.origin_user_id
+      elsif op_user_permissions.member?(:read_files)
+        permissions[:read] << token.origin_user_id
+      end
+    end
+
+    def project_folder_path(project)
+      "#{project.name} (#{project.id})".gsub(DISALLOWED_CHARS, '_').squish
+    end
+
+    def set_permissions(path, permissions)
+      Peripherals::Registry.resolve("commands.one_drive.set_permissions")
+                           .call(storage: @storage, path:, permissions:)
+                           .result_or do |error|
+        format_and_log_error(error, folder: project_folder_id.project_folder_id)
+      end
+    end
+
+    def rename_folder(source, target)
+      Peripherals::Registry
+        .resolve('commands.one_drive.rename_file')
+        .call(storage: @storage, source:, target:)
+        .result_or { |error| format_and_log_error(error, source:, target:) }
+    end
+
+    def create_folder(project_storage, folder_path)
+      Peripherals::Registry
+        .resolve('commands.one_drive.create_folder')
+        .call(storage: @storage, folder_path:)
+        .match(on_failure: ->(error) { format_and_log_error(error, folder_path:) },
+               on_success: ->(folder_info) do
+                 project_storage.update(project_folder_id: folder_info.id)
+                 project_storage.reload
+               end)
+    end
+
+    def remote_folders_map
+      using_admin_token do |http|
+        response = http.get("/v1.0/drives/#{@storage.drive_id}/root/children")
+
+        if response.status == 200
+          ServiceResult.success(result: filter_folders_from(response.json(symbolize_keys: true)))
+        else
+          errors = ::Storages::StorageError.new(code: response.status,
+                                                data: ::Storages::StorageErrorData.new(
+                                                  source: self.class, payload: response
+                                                ))
+          format_and_log_error(errors)
+          ServiceResult.failure(result: :error, errors:)
+        end
+      end
+    end
+
+    def filter_folders_from(json)
+      json.fetch(:value, []).each_with_object({}) do |item, hash|
```

Comment:
Not really. `each_with_object` is similar to `reduce` but different in that you explicitly pass in the accumulator object and it only mutates it if you do so.

Example:

In `reduce` (or `inject`) the last expression of the block, becomes the accumulator. So without the call to `acc` there, the next iteration would have been `10` instead of `{ 1 => 1 }`. IIRC there was some memory usage implication, but it might have changed.

```ruby
[1,2,3,4,5,6].reduce({}) { |acc, i| acc[i] = i ** i ;  acc }
```

`each_with_object` iterates through the collection passing the accumulator object back to the block each iteration. This means that if you don't make changes to the accumulator, nothing will happen as it ignores the result of the block, always returning the accumulator object.

```ruby
[1,2,3,4,5,6].each_with_object({}) {|i, acc| acc[i] = i ** i }
```

In general, unless I'm reducing to a single scalar, I prefer `each_with_object` as it is clearer on what is happening.

Reply:
Ah, brilliant explanation. then it boils down to a matter of choice. I'd always prefer reduce, as it is know over a variety of languages and hence easy to understand. The fact, that each_with_object does exactly the same, just without the need of returning the accumulator, smells a bit like "meh, I do not want to write that last line." and of course, it is very specific (make `reduce` with an object accumulator), while reduce can have any kind of accumulator.

Yet, I am fine with the usage, from reading the code, it is pretty obvious what it does.

Reply:
Maybe on second thought, I pretty not satisfied with the fact, that `each_with_object` is on design side effect based. While `reduce` explicitly displays in code, what happens with the aggregator.

Reply:
yes, that's the very issue of (im)mutability. `reduce` is working on immutable objects, while `each_with_object` mutates it.